source/common/router/router.h

```diff
@@ -163,6 +163,10 @@ class Filter : Logger::Loggable<Logger::Id::router>,
   }
   const Router::MetadataMatchCriteria* metadataMatchCriteria() const override {
     if (route_entry_) {
+      const auto impl = dynamic_cast<RequestInfo::RequestInfoImpl*>(&callbacks_->requestInfo());
```

Comment:
Instead of depending on the RequestInfoImpl here (and Router::MetadataMatchCriteriaImpl in the RequestInfoImpl), is there a reason you can't invoke `callbacks_->requestInfo().dynamicMetadata()` here and then construct the merged MetadataMatchCriteriaImpl in this function?

Reply:
I'm still not sure we should leak the match criteria into the RequestInfo interface. A new Router::Filter instance is created for each new stream, so I think it would be safe to track the Router::MetadataMatchCriteria in the filter.

Reply:
ah, i wasn't aware of that. let me change it then.

Reply:
ok, it's not that many places...